gen/field.go

```diff
@@ -94,23 +105,88 @@ func (f fieldGroupGenerator) DefineStruct(g Generator) error {
 			<end>
 		}`,
 		f,
-		TemplateFunc("tag", func(f *compile.FieldSpec) string {
-			// We want to add omitempty if the field is an optional struct or
-			// primitive to redure "null" noise. We won't add omitempty for
-			// optional collections because omitempty doesn't differentiate
-			// between nil and empty collections.
-
-			if (isStructType(f.Type) || isPrimitiveType(f.Type)) && !f.Required {
-				return fmt.Sprintf("`json:\"%s,omitempty\"`", f.Name)
-			}
-
-			return fmt.Sprintf("`json:%q`", f.Name)
-			// TODO(abg): Take go.tag and js.name annotations into account
-		}),
+		TemplateFunc("tag", generateTags),
 		TemplateFunc("declFieldName", f.declFieldName),
 	)
 }
 
+// generateTags parses the annotation on the thrift field and creates the resulting go tag
+func generateTags(f *compile.FieldSpec) string {
+	// start with default json tag
+	tk := []string{jsonTagName} // list of tag keys to preserve order
```

Comment:
yea I thought it'd be nice to preserve the author's order. I also think it helps readibility to see the json tag first since it will always be there. Other tags will be on some fields and not others so the tags wouldn't look nicely left-aligned since they could all have different starting tags